src/state_machine.zig

```diff
@@ -1527,6 +1558,55 @@ pub fn StateMachineType(
             return @sizeOf(Result(operation)) * count;
         }
 
+        /// - When `operation == .create_{accounts, transfers}`:
+        ///   It validates if the event's timestamp regressed in relation to the
+        ///   `commit_timestamp`.
+        ///   Panics in a past timestamp is found (except in aof_recovery mode).
+        ///
+        /// - When `operation == .import_{accounts, transfers}`:
+        ///   Allows past timestamps, but it validates if it regressed in relation
+        ///   to the last inserted object's timestamp.
+        ///   Panics if a future timestamp is found.
+        ///
+        /// This validation must be called _after_ the idempotency checks so the user can handle
+        /// `exists_*` results with the same behavior in both `create_*` and `import_*` operations.
+        /// Otherwise, `import_timestamp_must_not_regress` would have precedence over `exists` in
+        /// the case of retries with imported events. Also, idempotency checks don't include the
+        /// timestamp field, e.g. there is no "exists with different timestamp" result code.
+        fn validate_regressed_timestamp(
+            self: *const StateMachine,
+            comptime operation: Operation,
+            event: *const Event(operation),
+        ) bool {
+            switch (operation) {
+                .create_accounts, .create_transfers => {
+                    assert(event.timestamp > self.commit_timestamp or
+                        global_constants.aof_recovery);
+                    return true;
+                },
+                .import_accounts, .import_transfers => {
+                    // When importing past events with user-defined timestamp field, the "global"
+                    // timestamp cannot regress or repeat. It needs to be checked against the max
+                    // timestamp from both Accounts and Transfers.
+                    // If we ever have add a new groove keyed by timestamp, it must be included.
+                    const timestamp_max = @max(
+                        if (self.forest.grooves.accounts.objects.key_range) |*key_range|
+                            key_range.key_max
+                        else
+                            0,
+                        if (self.forest.grooves.transfers.objects.key_range) |*key_range|
+                            key_range.key_max
+                        else
+                            0,
+                    );
```

Comment:
> Could we treat the timestamp_max's as independent by object type?

I don't think we should do that, unless we want to relax the invariants  `t.timestamp > dr_account.timestamp and t.timestamp > cr_account.timestamp`.

Reply:
Ah, that's a good point.

Could we add a new error code for `import_transfers` to protect against that case? It might be worth it to make migration scripts simpler (unless I'm misunderstanding how they would be implemented?).